nx/lib/nx/random.ex

```diff
@@ -0,0 +1,374 @@
+    Nx.stack([y1, y2])
+    |> Nx.as_type({:u, 32})
+  end
+
+  defnp rolled_loop_step(i, {xs, ks, rs}) do
+    k1 = ks[0]
+    k2 = ks[1]
+    k3 = ks[2]
+    r1 = rs[0]
+    r2 = rs[1]
+
+    {_, xs, _} =
+      while {x = 0, xs, rs}, x < 4 do
+        {x + 1, apply_round(xs, rs[0][x]), rs}
+      end
+
+    xs1 = k1 + xs[0]
+
+    xs2 = k2 + i + 1 + xs[1]
+
+    new_xs =
+      Nx.stack([xs1, xs2])
+      |> Nx.as_type({:u, 32})
+
+    new_ks = Nx.stack([k2, k3, k1])
+    new_rs = Nx.stack([r2, r1])
+
+    {new_xs, new_ks, new_rs}
+  end
+
+  defnp rotate_left(x, rot) do
+    x <<< rot ||| x >>> (@nbits - rot)
+  end
+
+  defp float_info(type) do
+    case type do
+      {:bf, 16} -> [exp: 8, mantissa: 7]
+      {:f, 16} -> [exp: 5, mantissa: 11]
+      {:f, 32} -> [exp: 8, mantissa: 23]
+      {:f, 64} -> [exp: 11, mantissa: 52]
+    end
+  end
+
+  @doc """
+  Sample uniform random integer values in [min_val, max_val).
+
+  ## Options
+
+    * `:type` - an int type for the returned tensor
+
+    * `:shape` - shape of the returned tensor
+
+  """
+  defn randint(key, min_val, max_val, opts \\ []) do
+    opts = keyword!(opts, shape: {1}, type: {:s, 32})
+    assert_key(key)
+
+    shape = opts[:shape]
+    type = {_, nbits} = opts[:type]
+    case type do
+      {:u, _} -> :ok
+      {:s, _} -> :ok
+      _ -> raise ArgumentError,
+            "expected integer type, got type #{inspect(type)}"
+    end
+
+    keys = split(key)
+
+    min_val = Nx.broadcast(min_val, shape)
+    max_val = Nx.broadcast(max_val, shape)
+
+    higher_bits = random_bits(keys[0], shape: shape)
+    lower_bits = random_bits(keys[1], shape: shape)
+    span = max_val - min_val
+
+    multiplier =
+      Nx.power(2, Nx.quotient(nbits, 2))
+      |> Nx.remainder(span)
+      |> Nx.power(2)
+      |> Nx.remainder(span)
+
+    offset =
+      higher_bits
+      |> Nx.remainder(span)
+      |> Nx.multiply(multiplier)
+      |> Nx.add(Nx.remainder(lower_bits, span))
+      |> Nx.remainder(span)
+
+    Nx.as_type(min_val + offset, type)
+  end
+
+  @doc """
+  Sample uniform float values in [min_val, max_val).
+
+  ## Options
+
+    * `:type` - an float type for the returned tensor
+
+    * `:shape` - shape of the returned tensor
+
+    * `:min_val` - minimum value, default is 0.0
+
+    * `:max_val` - maximum value, default is 1.0
+
+  """
+  defn uniform(key, opts \\ []) do
+    opts = keyword!(opts, shape: {1}, type: {:f, 32}, min_val: 0.0, max_val: 1.0)
+
+    assert_key(key)
+
+    type = opts[:type]
+    case type do
+      {:f, _} -> :ok
+
+      _ -> raise ArgumentError,
+            "expected float type, got type #{inspect(type)}"
+    end
+
+    info = transform(opts[:type], &float_info(&1))
+
+    shape = opts[:shape]
+    type = {_dtype, nbits} = opts[:type]
+    min_val = opts[:min_val]
+    max_val = opts[:max_val]
+
+    u_one = Nx.tensor(1065353216, type: {:u, nbits})
+    #Equivalent of constant - Nx.tensor(1.0, type: type) |> Nx.bitcast({:u, nbits})

```

Comment:
The value `1065353216` will be different depending on the `type`, no? We can compute `u_one` at compile time, but it is worth the effort only if we can eliminate the other `bitcast` too.

Reply:
Yes, it is a 32bit placeholder until we figure out what to do with bitcast.

Reply:
I'll look into supporting bitcast from integer to float in Torchx, so don't worry about this :)

Reply:
I did an initial exploration of an alternative approach that doesn't rely on bit transformations, results in [this notebook](https://gist.github.com/jonatanklosko/462e0b2f0dbe06a42ae2c84224a59ba4). It seems that it could work, but could use someone with more expertise.

@polvalente regardless, I've just realised is that we use unsigned integers all across the implementation, so this alone kills torchx, right?

Sidenote: in `jax.random` [they say](https://github.com/google/jax/blob/07fcf793240645053987f918681da2c60c8c7103/jax/_src/random.py#L258-L261):

> The strategy here is to randomize only the mantissa bits with an exponent of
> 1 (after applying the bias), then shift and scale to the desired range. The
> bit-level transformation we use relies on Numpy and XLA having bit-for-bit
> equivalent float representations, which might not be true on all platforms.

Reply:
You're right! Let's not support torchx for now, then.

We can perhaps extend the code in a way that it operates the positive parte of the s64 domain in the future

Reply:
@jonatanklosko @polvalente Do you guys have any resources on integration with xla/torchx and possible caveats that we should be aware of in the future? To be honest as new to Nx as I am, I can't really follow your discussion without more knowledge.

Reply:
@VenomSolo no worries! Things should generally work fine with exla. As for torchx, the main differences is that u16/u32/u64 types are not supported, and there are a couple functions missing:

https://github.com/elixir-nx/nx/blob/8eadd775cf08b871c364794b231924170ff06e3d/torchx/lib/torchx/backend.ex#L1557-L1559

To verify you can configure either as a backend and try out some examples. In a script/livebook you can do

```elixir
Mix.install([
  {:nx, path: "~/path/to/nx/nx"},
  # We could also point to released versions, but since we have it locally
  # it's convenient to point there during development:
  {:exla, path: "~/path/to/nx/exla"},
  {:torchx, path: "~/path/to/nx/torchx"}
])

# Set the default backend for computing Nx operations
Nx.default_backend(EXLA.Backend)
# Nx.default_backend(Torchx.Backend)

key = Nx.Random.key(10)
Nx.Random.uniform(key)
```

Note that by default all operations are computed using `Nx.BinaryBackend`, which is a pure Elixir implementation.